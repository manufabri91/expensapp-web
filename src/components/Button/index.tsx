"use client";

interface Props {
  type: ButtonType;
  label: string;
  onClick?: () => void;
  ariaLabel?: string;
}

export enum ButtonType {
  Primary = "primary",
  Secondary = "secondary",
}

const buttonTypeClasses = new Map<ButtonType, string>([
  [
    ButtonType.Primary,
    "bg-blue-600 text-stone-100 hover:bg-blue-900 hover:text-stone-300 hover:border-stone-300 dark:bg-emerald-800 dark:border-stone-300 dark:text-stone-300 dark:hover:bg-teal-950",
  ],
  [
    ButtonType.Secondary,
    "bg-stone-100 text-slate-800 border-slate-800 hover:bg-stone-200 hover:text-slate-950 hover:border-slate-950 dark:bg-slate-800 dark:border-stone-100 dark:text-stone-100 dark:hover:bg-slate-900 dark:hover:border-stone-400 dark:hover:text-stone-400",
  ],
]);

export const Button = ({ type, label, onClick, ariaLabel }: Props) => {
  return (
    <button
      aria-label={ariaLabel || label}
      onClick={onClick}
      className={`px-4 py-2 rounded-full border-2 transition ease-in-out ${buttonTypeClasses.get(
        type
      )}`}
    >
      {label}
    </button>
  );
};
