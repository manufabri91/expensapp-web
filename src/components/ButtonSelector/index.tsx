interface RadioSelectorOption {
  id: string;
  label: string;
  colorClass?: string;
}

interface RadioSelectorProps {
  options: RadioSelectorOption[];
  value?: string;
  onChange: (id: string) => void;
  fieldName: string;
}

export const ButtonSelector = ({ options, value, onChange, fieldName }: RadioSelectorProps) => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    onChange(event.target.value);
  };
  return (
    <div className="flex w-full items-center justify-center">
      <div className="inline-flex w-full overflow-hidden rounded-lg border border-gray-300 dark:border-gray-600">
        {options.map((option) => (
          <label key={option.id} htmlFor={option.id} className="flex-1 cursor-pointer">
            <input
              type="radio"
              name={fieldName}
              id={option.id}
              value={option.id}
              className="peer sr-only"
              checked={value === option.id}
              onChange={handleChange}
            />
            <span className="peer-checked:border-brand-green-600 peer-checked:dark:border-brand-green-500 relative inline-flex size-full items-center justify-center space-x-2 py-2 pr-3 pl-7 text-sm peer-checked:border-b-2 peer-checked:bg-slate-300 hover:bg-slate-200 peer-checked:dark:bg-slate-800 hover:dark:bg-slate-600">
              <span
                className={`before:absolute before:top-[14px] before:left-3 before:size-2 before:rounded-full ${option.colorClass ?? 'before:bg-blue-500'}`}
              >
                {option.label}
              </span>
            </span>
          </label>
        ))}
      </div>
    </div>
  );
};
