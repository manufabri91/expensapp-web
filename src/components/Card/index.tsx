interface Props {
  title: string;
  children: React.ReactNode;
}

export const Card = ({ title, children }: Props) => {
  return (
    <div className="bg-white dark:bg-slate-800 p-4 rounded-lg shadow-md">
      <h3 className="text-lg font-bold text-gray-800  dark:text-gray-100">
        {title}
      </h3>
      {children}
    </div>
  );
};
