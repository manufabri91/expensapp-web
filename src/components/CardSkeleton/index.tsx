interface Props {
  className?: string;
}

export const CardSkeleton = ({ className }: Props) => {
  return (
    <div
      role="status"
      className={`max-w-sm animate-pulse rounded border border-gray-200 bg-white p-4 shadow md:p-6 dark:border-gray-700 dark:bg-slate-800 ${className}`}
    >
      <div className="mb-4 h-2.5 w-48 rounded-full bg-gray-200 dark:bg-slate-700"></div>
      <div className="mb-2.5 h-2 rounded-full bg-gray-200 dark:bg-gray-700"></div>
      <div className="h-2 rounded-full bg-gray-200 dark:bg-gray-700"></div>

      <span className="sr-only">Loading...</span>
    </div>
  );
};
