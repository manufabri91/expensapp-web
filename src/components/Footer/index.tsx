export const Footer = () => {
  return (
    <footer className=" p-4 text-center">
      <p>ManukoDEV</p>
      <p>2024 - All rights reserved.</p>
    </footer>
  );
};
