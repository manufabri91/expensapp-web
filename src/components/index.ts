export * from "@/components/Card";
export * from "@/components/Money";
export * from "@/components/Footer";
export * from "@/components/Navbar";
export * from "@/components/Button";
export * from "@/components/MenuButton";
export * from "@/components/StyledLink";