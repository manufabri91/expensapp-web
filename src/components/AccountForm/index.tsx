'use client';

import { Button, ButtonVariant } from '@/components/Button';
import { useToaster } from '@/components/Toast/ToastProvider';
import { useAccountForm } from '@/components/AccountForm/AccountFormProvider';
import { createAccount, editAccount } from '@/lib/actions/accounts';
import { useAccounts } from '@/lib/providers/AccountsProvider';
import { AccountResponse } from '@/types/dto';
import { data as currenciesList } from 'currency-codes';

import { Label, Modal, Select, TextInput } from 'flowbite-react';
import { useEffect, useState } from 'react';

export const AccountForm = () => {
  const { showToast } = useToaster();
  const { accountFormData, isOpen, closeAccountForm } = useAccountForm();
  const { addAccount } = useAccounts();
  const [createdAccount, setCreatedAccount] = useState<AccountResponse | null>(null);
  const [editedAccount, setEditedAccount] = useState<AccountResponse | null>(null);
  const [processing, setProcessing] = useState<boolean>(false);

  useEffect(() => {
    console.log(createdAccount);
    if (createdAccount) {
      showToast(`Created Account with ID: ${createdAccount.id}`);
      setCreatedAccount(null);
      closeAccountForm();
      setProcessing(false);
    } else if (editedAccount) {
      showToast(`Edited Account with ID: ${editedAccount.id}`);
      setEditedAccount(null);
      closeAccountForm();
      setProcessing(false);
    }
  }, [closeAccountForm, createdAccount, editedAccount, showToast]);

  const submitHandler = async (formData: FormData) => {
    setProcessing(true);
    if (!accountFormData) {
      const createdAccount = await createAccount(formData);
      setCreatedAccount(createdAccount);
      addAccount(createdAccount);
    } else {
      const updatedAccocreatedAccount = await editAccount(formData);
      setEditedAccount(updatedAccocreatedAccount);
    }
  };

  if (!isOpen) return null;

  return (
    <Modal show={isOpen} size="md" onClose={closeAccountForm} popup>
      <Modal.Header as={'div'} className="m-4">
        <h3>{accountFormData ? 'Edit ' : 'Create '}Account</h3>
      </Modal.Header>
      <Modal.Body>
        <form className="flex max-w-md flex-col gap-4" action={submitHandler}>
          {!!accountFormData && (
            <div className="hidden">
              <div className="mb-2 block">
                <Label htmlFor="id" value="ID" />
              </div>
              <TextInput id="id" name="id" type="text" value={accountFormData?.id} readOnly shadow />
            </div>
          )}
          <div>
            <div className="mb-2 block">
              <Label htmlFor="name" value="Account Name" />
            </div>
            <TextInput id="name" name="name" type="text" defaultValue={accountFormData?.name} required shadow />
          </div>
          {!accountFormData && (
            <div>
              <div className="mb-2 block">
                <Label htmlFor="initialBalance" value="Initial Balance" />
              </div>
              <TextInput
                id="initialBalance"
                name="initialBalance"
                type="number"
                defaultValue={0}
                min="0"
                step=".01"
                required
                shadow
              />
            </div>
          )}

          <div className="max-w-md">
            <div className="mb-2 block">
              <Label htmlFor="currency" value="Currency" />
            </div>
            <Select id="currency" name="currency" defaultValue={accountFormData?.currency} required>
              <option value={undefined}>Select Currency</option>
              {currenciesList.map((currency) => (
                <option key={currency.code} value={currency.code}>
                  {currency.code} - {currency.currency}
                </option>
              ))}
            </Select>
          </div>

          <Button type="submit" variant={ButtonVariant.Primary} isProcessing={processing}>
            {accountFormData ? 'Edit' : 'Create'}
          </Button>
        </form>
      </Modal.Body>
    </Modal>
  );
};
