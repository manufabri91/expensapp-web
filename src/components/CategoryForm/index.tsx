'use client';

import { useEffect, useState } from 'react';
import { Label, Modal, TextInput } from 'flowbite-react';

import { Button, ButtonVariant } from '@/components/Button';
import { useToaster } from '@/components/Toast/ToastProvider';
import { CategoryResponse } from '@/types/dto';
import { useCategoryForm } from '@/components/CategoryForm/CategoryFormProvider';
import { useCategories } from '@/lib/providers/CategoriesProvider';
import { createCategory, editCategory } from '@/lib/actions/categories';
import { IconPickerFormField } from '@/components/IconPicker';

export const CategoryForm = () => {
  const { showToast } = useToaster();
  const { categoryFormData, isOpen, closeCategoryForm } = useCategoryForm();
  const { addCategory } = useCategories();
  const [createdCategory, setCreatedCategory] = useState<CategoryResponse | null>(null);
  const [editedCategory, setEditedCategory] = useState<CategoryResponse | null>(null);
  const [processing, setProcessing] = useState<boolean>(false);

  useEffect(() => {
    if (createdCategory) {
      showToast(`Created Category with ID: ${createdCategory.id}`);
      setCreatedCategory(null);
      closeCategoryForm();
      setProcessing(false);
    } else if (editedCategory) {
      showToast(`Edited Category with ID: ${editedCategory.id}`);
      setEditedCategory(null);
      closeCategoryForm();
      setProcessing(false);
    }
  }, [closeCategoryForm, createdCategory, editedCategory, showToast]);

  const submitHandler = async (formData: FormData) => {
    setProcessing(true);
    if (!categoryFormData) {
      const createdCategory = await createCategory(formData);
      setCreatedCategory(createdCategory);
      addCategory(createdCategory);
    } else {
      const updatedAccocreatedCategory = await editCategory(formData);
      setEditedCategory(updatedAccocreatedCategory);
    }
  };

  if (!isOpen) return null;

  return (
    <Modal show={isOpen} size="2xl" onClose={closeCategoryForm} popup>
      <Modal.Header as={'div'} className="m-4">
        <h3>{categoryFormData ? 'Edit ' : 'Create '}Category</h3>
      </Modal.Header>
      <Modal.Body>
        <form className="flex flex-col gap-4" action={submitHandler}>
          {!!categoryFormData && (
            <div className="hidden">
              <div className="mb-2 block">
                <Label htmlFor="id" value="ID" />
              </div>
              <TextInput id="id" name="id" type="text" value={categoryFormData?.id} readOnly shadow />
            </div>
          )}
          <div>
            <div className="mb-2 block">
              <Label htmlFor="name" value="Category Name" />
            </div>
            <TextInput
              addon={
                <IconPickerFormField
                  id="iconName"
                  name="iconName"
                  initialValue={categoryFormData?.iconName}
                ></IconPickerFormField>
              }
              id="name"
              name="name"
              type="text"
              defaultValue={categoryFormData?.name}
              required
              shadow
            />
          </div>
          <div>
            <div className="mb-2 block">
              <Label htmlFor="color" value="Category Color" />
            </div>
            <TextInput id="color" name="color" type="text" defaultValue={categoryFormData?.color ?? ''} shadow />
          </div>

          <Button type="submit" variant={ButtonVariant.Primary} isProcessing={processing}>
            {categoryFormData ? 'Edit' : 'Create'}
          </Button>
        </form>
      </Modal.Body>
    </Modal>
  );
};
