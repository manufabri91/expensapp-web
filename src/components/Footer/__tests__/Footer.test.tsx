import { render, screen } from '@testing-library/react';
import React from 'react';
import { Footer } from '@/components/Footer';

describe('Footer component', () => {
  it('should render correctly', () => {
    const { container } = render(<Footer />);

    expect(container).toMatchSnapshot();
  });

  it('should render the company name', () => {
    render(<Footer />);
    expect(screen.getByText('ManukoDEV')).toBeInTheDocument();
  });

  it('should render the rights reserved text', () => {
    render(<Footer />);
    expect(screen.getByText('2024 - All rights reserved')).toBeInTheDocument();
  });

  it('should have the correct class names', () => {
    const { container } = render(<Footer />);
    const footerDiv = container.firstChild;
    expect(footerDiv).toHaveClass('container');
    expect(footerDiv?.firstChild).toHaveClass('companyName');
    expect(footerDiv?.lastChild).toHaveClass('text');
  });
});
