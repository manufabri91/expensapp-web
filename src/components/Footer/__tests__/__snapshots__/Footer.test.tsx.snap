// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component should render correctly 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="companyName"
    >
      ManukoDEV
    </div>
    <div
      class="text"
    >
      2024 - All rights reserved
    </div>
  </div>
</div>
`;
